RFC - Allow routes to restrict the hosts they should match on

markstory opened the pull request:
By passing the request we can make it possible to add route conditions for more than just the URL. I've had to add new methods onto Router and RouteCollection to make this possible, but I feel that passing the object around is better long term anyways.

This requires deprecating `Router::parse()` which makes me a bit sad, but I feel this makes the code simpler/better long term.

This is currently an RFC as I want to see if people think the addition of the `_host` option is worth the churn in code. Also if there are any other similar types of conditions/matchers we should add.

Refs #9917

markstory committed:
First pass at making it possible to add host conditions.

By passing the request we can make it possible to add route conditions
for more than just the URL. I've had to add new methods onto Router and
RouteCollection to make this possible, but I feel that passing the
object around is better long term anyways.

markstory committed:
Use new Request based parsing in Middleware and Filter.

markstory committed:
Fix undefined variables.

markstory committed:
Add tests for RouteCollection::parseRequest()

Add test for non-host matching parts of parseRequest()

markstory committed:
Add missing doc block and stub implementation.

markstory committed:
Move tests around.

markstory committed:
Only assign $host once.

markstory committed:
Add host name matching to RouteCollection/Route

Add tests at RouteCollection layer as it is simpler/more holistic place
to test the behavior.

markstory committed:
Add docs for _host option.

markstory committed:
Wrap long lines.

markstory committed:
Fix PHPCS errors.

ADmad commented on src/Routing/RouteCollection.php:
```diff
@@ -148,6 +149,45 @@ public function parse($url, $method = '')
+
+            /* @var \Cake\Routing\Route\Route $route */
+            foreach ($this->_paths[$path] as $route) {
+                if (!$route->hostMatches($host)) {
```
Why is the host check done here instead of inside `Route::parse()` where the method type and other checks are done?

ADmad replied:
Answering myself, it is because `Route::parse()` does not have the host name info :)

markstory replied:
That and if the host doesn't match there is no need in running a possibly expensive route match operation.

ADmad replied:
Well you would obviously return early if the hosts didn't match :) I was just concerned that only one specific option was being checked out of Route::parse() but I don't think it's a problem.

markstory replied:
I could add a new `parse` like method to the route if you think the logic is out of place here.

ADmad replied:
I would prefer if checks for all options were initiated from same method by it's your call.

chinpei215 commented on src/Routing/Middleware/RoutingMiddleware.php:
```diff
@@ -45,10 +45,7 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
-                        $request->getUri()->getPath(),
-                        $request->getMethod()
-                    )
+                    Router::parseRequest($request)
```
Wouldn't this change break any existing custom route classes?

markstory replied:
I wanted to keep the deprecations small, so only `Router` and `RouteCollection` _needed_ new methods.

lorenzo commented:
I like the direction this is going, but I'm worried that custom route classes don't have access to the request as well. What if we create a new Interface the route classes can implement to hint they are ready to do `parseRequest()` ?

markstory commented:
> What if we create a new Interface the route classes can implement to hint they are ready to do `parseRequest()`

I can add `parseRequest()` to `Route`, and deprecate the `parse()` method.

markstory committed:
Push host checking into Route.

Add a `parseRequest` method to `Route` as well. This gives the 'new' API
a consistent implementation across all the routing layers. It also
provides a more flexible way for people to customize routing as they
have access to the entire request via the new API.

markstory commented:
@lorenzo Updated with `Route::parseRequest()` now. It is a nicer looking implementation now, thanks for the suggestion folks :+1:

markstory committed:
Fix PHPCS errors.

josegonzalez commented:
I like it. Is it any slower?

markstory commented:
>  Is it any slower?

That is a great question. I don't know yet, but I'll find out.

lorenzo commented:
Great job @markstory!

I think it is still interesting to know the performance impact of the changes, but in any case the benefits outweigh any possible (small) performance hit

markstory commented:
I wrote a [very simple test script](https://gist.github.com/markstory/31d4ca1baffb01e9638a5f00bbdbfad7) and these were the results:

``` 
$ php routes.php
Router::parse() 0.060201

Router::parseRequest() 0.065646

$ php routes.php
Router::parse() 0.060170

Router::parseRequest() 0.058235

$ php routes.php
Router::parse() 0.062613

Router::parseRequest() 0.062625

$ php routes.php
Router::parse() 0.061117

Router::parseRequest() 0.065163

$ php routes.php
Router::parse() 0.067012

Router::parseRequest() 0.074516

$ php routes.php
Router::parse() 0.058143

Router::parseRequest() 0.062988
```

It looks like a wash as far as performance is concerned.

josegonzalez commented:
Seems reasonable to me :)
